### ISSUE TITLE
[question] cores reported as sys.maxsize

### ISSUE THREAD

**Issue by vsoch** [snakemake/snakemake#213]
This is more of an observation that I had, and I'm just walking through the code to see how it generates. Specifically, when we use a flavor of cloud cluster executor, the number of cores is reported to be an unusually large number. It turns out it's the sys.maxsize:

```python
    if cluster or cluster_sync or drmaa or tibanna or google_lifesciences:
        cores = sys.maxsize
    else:
        nodes = sys.maxsize
```
I'm not sure if this is done intentionally, but if it is, since we print it out as cores to the user, there might be some better way to go about this. I'm opening the issue for now for discussion, I know @moschetti had asked about it. I'll update the issue with an example of the printing if/when I stumble on one (very likely will!) during testing stuffs.

**Comment by vsoch** [comment]
As promised, here is what it looks like to the user running the pipeline:

```
Using shell: /bin/bash
Provided cores: 9223372036854775807
Rules claiming more threads will be scaled down.
Job counts:
	count	jobs
	1	all
	1	build_oncoprint_table
	1	plot_oncoprint
	3
```
so many cores!!

**Comment by moschetti** [comment]
TL;DR - sys.maxsize doesn't count size of system, but reflects the max pointer size, use os.cpu_count() instead

That number is 2**63 - 1
Python's [sys.maxsize](https://docs.python.org/3/library/sys.html#sys.maxsize) lists the maximum value that can be stored in the underlying C compiler's ssize_t variable, so it should report that value on all 64-bit systems (and likewise report 2**31 - 1 on a 32-bit system. It shouldn't be used to count cores.

[os.cpu_count()](https://docs.python.org/3/library/os.html#os.cpu_count) will report the number of cores on the system (as of Python >= 3.4), or len(os.sched_getaffinity(0)) to report the number of usable cores.

So [the code that is setting that cores/nodes](https://github.com/snakemake/snakemake/blob/0607695047290effb44367cd004523e5e3398171/snakemake/__init__.py#L321) should be updated.

The other uses of it ([1](https://github.com/snakemake/snakemake/blob/5bb0ed743c8850532a8d4c485105fefc9cf33bc5/snakemake/report/__init__.py#L310), [2](https://github.com/snakemake/snakemake/blob/e064a076ebb01310447ef9d09b229951cf52a04a/snakemake/jobs.py#L79)) are using it to represent an upper bound. Although in Python 3.x,` int` is unbounded so this isn't strictly necessary anymore, but it seems fine to set an upper limit at a 64-bit int since you're picking some high number anyway.

**Comment by vsoch** [comment]
@johanneskoester should we update to use `os.cpu_count()` ? For the remote executors, it won't matter so much since the host is just deploying the job. But if that cpu count variable is used as some default to choose the machine type, perhaps instead it should be set to some reasonable number like 4?

**Comment by linsalrob** [comment]
I have exactly this issue. With a trivial snakefile:

```
print(f"There are {workflow.cores} cores available")
```

If I run on the same machine:

```
snakemake -s test.snakefile  --cores 10
```

I get 

```
There are 10 cores available
```

If I run:

```
snakemake -s test.snakefile  --cores 10 --cluster 'qsub'
```

I get:
```
There are 9223372036854775807 cores available
```

My problem is I want to do split input files based on how many cores I am told that I have available, so I can't do it with this approach. Any pointers?

**Comment by vsoch** [comment]
@linsalrob I can put in a PR with a quick fix first thing tomorrow, and then you would have something to test and we would be on our way! Does that sound like a good plan?

**Comment by linsalrob** [comment]
Yes, for sure. Happy to test and a great plan!

**Comment by vsoch** [comment]
Great! I’ll ping you here tomorrow. Have a good night!

**Comment by vsoch** [comment]
@linsalrob here you go!

https://github.com/snakemake/snakemake/pull/346#issuecomment-618337800

The tiniest PR ever. But small and mighty! Give it a test and post an update (if you are happy with the fix) in the PR!

**Comment by vsoch** [comment]
One question I’m curious about is why Snakemake doesn’t honor the user specification of cores given cluster, Tibanna, etc. In your second example above, with my change it would still just use os.cpu_count(). I want to suggest that we use args cores OR the os.cpu_count(), but want to bring it up for discussion here because I suspect there is a reason it’s implemented like this. @johanneskoester can you comment? I can update the PR.

**Comment by linsalrob** [comment]
Yes, I agree with @vsoch that `os.cpu_count()` is not the right solution. I dug around some more, and I think the change should be:

```
min(workflow.cores, workflow.nodes)
```

From my previous example, changing the print statement to be:

```
print(f"There are {workflow.cores} cores and {workflow.nodes} nodes available")
```

Produces these two outputs:

without `--cluster`:
```
snakemake -s test.snakefile  --cores 10 
There are 10 cores and 9223372036854775807 nodes available
```

with `--cluster`
```
$ snakemake -s test.snakefile  --cores 10 --cluster 'qsub'
There are 9223372036854775807 cores and 10 nodes available
```

It is a bit confusing since `--cluster` still uses the `--cores` option but stores it as `workflow.nodes`!

**Comment by vsoch** [comment]
Cool - thanks for the feedback! Let's wait for @johanneskoester to weigh in about what he would like to see.

**Comment by johanneskoester** [comment]
To be honest, it is a hack. The scheduling algo needs a value for cores, and I don't want to constrain the scheduling by the number of cores when running in cluster mode. 

@linsalrob your observation comes from the fact that --cores is actually an overloaded argument. It is at the same time --jobs, which is the alias that should be used in combination with --cluster.

Instead of setting cores to sys.maxsize, I think it is better to set it to None, and just set it locally in the scheduler.py to sys.maxsize (which is needed for the scheduling algorithm).

**Comment by vsoch** [comment]
@johanneskoester I can update the PR to do that.

@linsalrob it seems like you might want to try using --jobs then? I also think that the output is confusing to see ("Wow, I didn't know I had that many cores!") so we might want to avoid the print if the original value is set to None from the entrypoint to snakemake (indicating cluster mode).

**Comment by vsoch** [comment]
@johanneskoester can you point me to where I need to make changes? I just spent some time looking over the files and I'm not 100% sure which executors this is relevant for, and more importantly, where in the files I need to set sys.maxsize, None, and then to silence the print.  I can update my PR if you can easily direct me to these spots, but I think it might be quicker if you are able to do it!

**Comment by johanneskoester** [comment]
Should be fixed in PR #353.

**Comment by vsoch** [comment]
I think this was fixed with #353 ? Closing, and please open a new issue / re-open and update this one if there is still something to work on.
